VirtusizeAuth/Sources/VirtusizeAuthorization.swift

```diff
@@ -0,0 +1,269 @@
+			str: authUrlString,
+			// start with `&` delimiter to ensure `fallback_redirect_uri` is not detect instead
+			after: "&\(Constants.queryRedirectUriKey)=",
+			before: "&",
+			default: ""
+		)
+		let region = substringBetweenDelimiters(
+			str: originalRedirectUri,
+			after: "virtusize.",
+			before: "/",
+			default: "com"
+		)
+		let env = substringBetweenDelimiters(
+			str: originalRedirectUri,
+			after: "sns-proxy/",
+			before: "/",
+			default: "staging"
+		)
+
+		if !originalRedirectUri.isEmpty {
+			authUrlString = authUrlString.replacingOccurrences(
+				of: originalRedirectUri,
+				with: VirtusizeAuthStringHelper.getRedirectUrl(region: region, env: env)
+			)
+		}
+
+		// HACK `channel_url` breaks the Facebook OAuth, it seems obsolete, so we remove it from the URL
+		authUrlString = VirtusizeAuthStringHelper.removeParamFromUrl(
+			url: authUrlString,
+			paramName: "channel_url"
+		)
+
+		authUrlString = updateUrlWithAppBundleId(urlString: authUrlString)
+		authUrlString = updateUrlWithSnsType(urlString: authUrlString)
+		authUrlString = updateUrlWithEnvironment(urlString: authUrlString, region: region, env: env)
+
+		VirtusizeLogger.debug("AuthUrl: \(authUrlString)")
+
+		if let authURL = URL(string: authUrlString) {
+			let virtusizeSVC = VirtusizeSafariViewController(webView: webView, url: authURL, onClose: { [weak self] in
+				// reset reference to the controller
+				self?.safariVewController = nil
+			})
+			safariVewController = virtusizeSVC
+			viewController.present(virtusizeSVC, animated: true)
+			return true
+		}
+
+		return false
+	}
+
+	/// Updates the auth url with the environment params appended to the state query parameter.
+	/// Those parameters are important to restore redirectUri when returned to the app by callback.
+	internal func updateUrlWithEnvironment(
+		urlString: String,
+		region: String,
+		env: String
+	) -> String {
+
+		let url = updateUrlStateWithProperty(
+			urlString: urlString,
+			key: VirtusizeAuthConstants.regionKey,
+			value: region)
+
+		return updateUrlStateWithProperty(
+			urlString: url,
+			key: VirtusizeAuthConstants.envKey,
+			value: env)
+	}
+
+	/// Updates the auth url with the app bundle id appended to the state query parameter.
+	private func updateUrlWithAppBundleId(
+		urlString: String
+	) -> String {
+		guard let bundleId = Bundle.main.bundleIdentifier else {
+			return urlString
+		}
+
+		return updateUrlStateWithProperty(
+			urlString: urlString,
+			key: Constants.nativeAppIdKey,
+			value: bundleId)
+	}
+
+	/// Updates the auth url with the app bundle id appended to the state query parameter.
+	internal func updateUrlWithSnsType(
+		urlString: String
+	) -> String {
+		let snsType: SNSType? =
+			if urlString.contains("facebook") {
+				SNSType.facebook
+			} else if urlString.contains("google") {
+				SNSType.google
+			} else if urlString.contains("line.me") {
+				SNSType.line
+			} else {
+				nil
+			}
```

Comment:
```Swift
enum SNSType: String {
    case facebook
    case google
    case line
    
    static func from(url: String) -> SNSType? {
        let patterns: [SNSType: String] = [
            .facebook: "facebook",
            .google: "google",
            .line: "line.me"
        ]
        
        return patterns.first(where: { url.contains($0.value) })?.key
    }
}

let snsType = SNSType.from(url: urlString)
```
What about adding a function to the SNSType enum? It might make it easier to manage, especially if we need to add or remove SNS types in the future.

Reply:
done